moveit_ros/perception/handeye_calibration_target/src/handeye_aruco_target.cpp

```diff
@@ -0,0 +1,301 @@
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Yu Yan */
+
+#include <moveit/handeye_calibration_target/handeye_aruco_target.h>
+
+namespace moveit_handeye_calibration
+{
+const std::string LOGNAME = "handeye_aruco_target";
+
+bool HandEyeArucoTarget::initialize(const int& markers_x, const int& markers_y, const int& marker_size,
+                                    const int& separation, const int& border_bits, const std::string& dictionary_id,
+                                    const double& marker_measured_size, const double& marker_measured_separation)
+{
+  marker_dictionaries_ = { { "DICT_4X4_250", cv::aruco::DICT_4X4_250 },
+                           { "DICT_5X5_250", cv::aruco::DICT_5X5_250 },
+                           { "DICT_6X6_250", cv::aruco::DICT_6X6_250 },
+                           { "DICT_7X7_250", cv::aruco::DICT_7X7_250 },
+                           { "DICT_ARUCO_ORIGINAL", cv::aruco::DICT_ARUCO_ORIGINAL } };
+
+  target_params_ready_ =
+      setTargetIntrinsicParams(markers_x, markers_y, marker_size, separation, border_bits, dictionary_id) &&
+      setTargetDimension(marker_measured_size, marker_measured_separation);
+
+  return target_params_ready_;
+}
+
+std::vector<std::string> HandEyeArucoTarget::getDictionaryIds()
+{
+  std::vector<std::string> dictionary_ids;
+  for (const std::pair<const std::string, cv::aruco::PREDEFINED_DICTIONARY_NAME>& name : marker_dictionaries_)
+    dictionary_ids.push_back(name.first);
+  return dictionary_ids;
+}
+
+bool HandEyeArucoTarget::setTargetIntrinsicParams(const int& markers_x, const int& markers_y, const int& marker_size,
+                                                  const int& separation, const int& border_bits,
+                                                  const std::string& dictionary_id)
+{
+  if (markers_x != markers_x_ || markers_y != markers_y_ || marker_size != marker_size_ || separation != separation_ ||
+      border_bits != border_bits_ || marker_dictionaries_.find(dictionary_id)->second != dictionary_id_)
```

Comment:
What if the marker_dictionaries_ does not contain dictionary_id? Wouldn't that cause issues here as we expect the find method to always return a valid iterator.

Reply:
This block was used previously debug print the input parameters when any value of them is different in the new one. A duplicate check on the existence of dictionary_id happens later in the code to make a response. I refactored this [function](https://github.com/ros-planning/moveit/pull/1558/files#diff-78064d7ec7c6ae7b035c309222f83192R73-R100) to correct the duplicate check.